File: java/course.java
public class Course {
	private	String name;
	private String code;
	private int max_;
	private int min_;
	private	ArrayList<Professor> professors;
	private ArrayList<Enroll> enrollments; 

	Course(String name, String code, int max, int min, ArrayList<Professor> professor) {
		this.name = name;
		this.code = code;
		this.max_ = max;
		this.min_ = min;
		this.professors = [];
		this.enrollments = [];

		if (professor instanceof Professor) {
			this.professors.add(professor);
		} else if (professor instanceof ArrayList<>) {
			for (int i = 0; i < professor.length(); i++) {
				if (!(professor[i] instanceof Professor)) {
					throw new Exception("Invalid professor ...");
				}
			}
			self.professors = professor;
		} else {
			throw new Exception("Invalid Professor ...");
		}
	}

	public void addProfessor(ArrayList<Professor> professor) {
		if (!(professor instanceof Professor)) {
			throw new Exception("Invalid professor ...");
		}
		
		this.professors.add(professor);
	}

	public void addEnrollment(ArrayList<Enroll> enroll) {
		if (!(enroll instanceof Enroll)) {
			throw new Exception("Invalid enroll");
		}

		if (enrollments.length() == this.max_) {
			throw new Exception("Cannot enroll, course is full ...");
		}

		this.enrollments.add(enroll);
	}

	public boolean isCancelled() {
		return this.enrollments.length() < this.min_;
	}
	
}

File: javascript/course.js
const Professor = require('./professor');
const Enroll = require('./enroll');

class Course {
	constructor(name, code, max_, min_, professor) {
		this.name = name;
		this.code = code;
		this.max = max_;
		this.min = min_;
		this.professors = [];
		this.enrollments = [];

		if (professor instanceof Professor) {
			this.professors.push(professor);
		} else if (professor.isArray()) {
			professor.forEach(item => {
				if (!(item instanceof Professor)) {
					throw Error('Invalid Professor');
				}
			});
			this.professors = professor;
		} else {
			throw Error('Invalid Professor...');
		}
	}

	addProfessor(professor) {
		if (!(professor instanceof Professor)) {
			throw Error("Invalid Professor...")
		}
		this.professors.push(professor)
	}

	addEnrollment(enroll) {
		if (!(enroll instanceof Enroll)) {
			throw Error("Invalid Enroll");
		}
		if (this.enrollments.length === this.max) {
			throw Error("Cannot enroll, course is full ....");
		}

		this.enrollments.push(enroll);
	}

	isCancelled() {
		return this.enrollments.length < this.min;
	}
}

module.exports = Course;